randovania/game_connection/connector/am2r_remote_connector.py

```diff
@@ -0,0 +1,241 @@
+                continue
+            locations.add(PickupIndex(int(index)))
+
+        for location in locations:
+            self.PickupIndexCollected.emit(location)
+
+    def new_inventory_received(self, new_inventory: str) -> None:
+        """
+        __New inventory implementation detail:__
+        The game keeps track of which items it collected in a big string with the format
+        `items:XA|XB,YA|YB,ZA|ZB...`, where XA/YA/ZA are the item names and XB/YB/ZB is the quantity of that item.
+        This string is also saved to the save file. The item names *mostly* match up with the RDV item names,
+        except for ammo and progressives.
+        The game sends that string to RDV, where we then loop through it. We adjust the item name for ammo and
+        progressives accordingly, and stuff it into an item|quantity dict.
+        If the item exists in that dict, we increase its quantity, if not, we add it. Then, based on that dict, we
+        update the inventory.
+
+        Regarding ammo and progressives: for some reason, we need here the name of the ammo, and not the name of the
+        expansions. I.e. for `Missile Expansions`, we need `Missiles`. So there's a lookup dict here to adjust for them.
+
+        :param new_inventory: A string from the game with the format `items:XA|XB,YA|YB,ZA|ZB...`.
+        """
+        locations: dict[ItemResourceInfo, int] = {}
+        start_of_inventory = "items:"
+
+        if not new_inventory.startswith(start_of_inventory):
+            self.logger.warning("Unknown response: %s", new_inventory)
+            return
+
+        for position in new_inventory[len(start_of_inventory) :].split(","):
+            print(bytes(position, "utf-8"))
+            if "|" not in position:
+                self.logger.warning("Response should contain a '|', but it doesn't")
+                continue
+            (item_name, quantity) = position.split("|")
+            if not quantity.isdigit():
+                self.logger.warning("Response should contain a digit, but instead contains '%s'", position)
+                continue
+
+            # Ammo and progressive stuff is special, and this is my best idea on how to circumvent them :(
+            item_name_replacement = {
+                "Missile Expansion": "Missiles",
+                "Super Missile Expansion": "Super Missiles",
+                "Power Bomb Expansion": "Power Bombs",
+            }
+
+            progressives = {
+                "Progressive Jump": ("Hi-Jump Boots", "Space Jump"),
+                "Progressive Suit": ("Varia Suit", "Gravity Suit"),
+            }
+
+            if item_name in item_name_replacement:
+                item_name = item_name_replacement[item_name]
+
+            if item_name in progressives:
+                if len([i for i, q in locations.items() if i.long_name == progressives[item_name][0]]) > 0:
+                    item_name = progressives[item_name][1]
```

Comment:
Is this really `item_name = progressives[item_name][1]` with index 1 and not 0?

Reply:
Yeah. So basically with progressive jumps:
if our temp inventory has hijump collected, then we add space jump
if our temp inventory has space jump collected, then we add space jump
if our temp inventory has nothing collected, then we add hijump


I should probably reorder these to be a bit more logical though.

Reply:
Still don't know if I understood this code correctly. So far, I don't like that you iterating over the same thing you are actually building with this loop.
This means, you have the implict requirement of all items being processed by the loop before theit respective progressive e.g. Hi-jump must always be added by this loop before Progressive jump.

Reply:
>  e.g. Hi-jump must always be added by this loop before Progressive jump.

Yes. `Progressive Jump` will never be added. Progressives are not resources in the database. So I basically check how many progressives the game has gotten, and then add the actual resources to our list that we're building.